cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1138,6 +1160,22 @@ struct GroupedProductImpl final
     return Status::OK();
   }
 
+  static Status Consume(const Scalar& value, const int64_t count, c_type* reduced,
+                        int64_t* counts, uint8_t* no_nulls, const uint32_t* g) {
+    if (value.is_valid) {
+      const auto v = to_unsigned(static_cast<c_type>(UnboxScalar<Type>::Unbox(value)));
+      for (int i = 0; i < count; i++) {
+        reduced[*g] = static_cast<c_type>(to_unsigned(reduced[*g]) * v);
```

Comment:
Why doesn't this use `MultiplyTraits<AccType>::Multiply` instead? It seems like copy/pasting is putting us at risk of latent bugs here. Perhaps you can factor out the raw computation routines?
```c++
using InputCType = typename TypeTraits<Type>::CType;

AccType Reduce(const DataType& type, c_type u, InputCType v) {
  return MultiplyTraits<AccType>::Multiply(type, u, static_cast<c_type>(v));
}

AccType Reduce(const DataType& type, c_type u, c_type v) {
  return MultiplyTraits<AccType>::Multiply(type, u, v);
}
```

Same for summing:
```c++
using InputCType = typename TypeTraits<Type>::CType;

AccType Reduce(const DataType& type, c_type u, InputCType v) {
  return static_cast<c_type>(to_unsigned(u) + to_unsigned(static_cast<c_type>(v)));
}

AccType Reduce(const DataType& type, c_type u, c_type v) {
  return static_cast<c_type>(to_unsigned(u) + to_unsigned(v));
}
```

Then you may even reconcile the Sum and Product implementations further.

Reply:
Yes, this is a bad merge on my part - I will fix.